src/v/storage/readers_cache.cc

```diff
@@ -0,0 +1,252 @@
+
+#include <chrono>
+
+namespace storage {
+
+readers_cache::readers_cache(
+  model::ntp ntp, std::chrono::milliseconds eviction_timeout)
+  : _ntp(std::move(ntp))
+  , _eviction_timeout(eviction_timeout) {
+    // setup eviction timer
+    _eviction_timer.set_callback([this] {
+        (void)ss::with_gate(_gate, [this] {
+            return maybe_evict().finally([this] {
+                if (!_gate.is_closed()) {
+                    _eviction_timer.arm(_eviction_timeout);
+                }
+            });
+        }).handle_exception_type([](const ss::gate_closed_exception&) {
+            // ignore, we are stopping
+        });
+    });
+
+    _eviction_timer.arm(_eviction_timeout);
+}
+
+model::record_batch_reader
+readers_cache::put(std::unique_ptr<log_reader> reader) {
+    if (reader->lease_range_base_offset() < model::offset{0}) {
+        // do not cache reader with empty lease
+        return model::record_batch_reader(std::move(reader));
+    }
+    vlog(
+      stlog.trace,
+      "{} - adding reader [{},{}]",
+      _ntp,
+      reader->lease_range_base_offset(),
+      reader->lease_range_end_offset());
+
+    auto ptr = new entry{.reader = std::move(reader)}; // NOLINT
+    _readers.push_back(*ptr);
+    return ptr->make_cached_reader(this);
+}
+
+std::optional<model::record_batch_reader>
+readers_cache::get_reader(const log_reader_config& cfg) {
+    vlog(stlog.trace, "{} - trying to get reader for: {}", _ntp, cfg);
+    intrusive_list<entry, &entry::_hook> to_evict;
+    /**
+     * We use linear search since _readers intrusive list is small.
+     */
+    auto it = _readers.begin();
+    while (it != _readers.end()) {
+        const auto is_valid = it->reader->is_reusable() && it->valid;
+        // TODO: think about relaxing this equality condition
+        const auto offset_matches = it->reader->next_read_lower_bound()
+                                    == cfg.start_offset;
+        // if invalid we will dispose this entry in background
+        if (!is_valid) {
+            it = _readers.erase_and_dispose(
+              it, [&to_evict](entry* e) { to_evict.push_back(*e); });
+            continue;
+        }
+        if (offset_matches && is_valid) {
+            // found matching reader
+            break;
+        }
+        ++it;
+    }
+    /**
+     * dispose unused readers in background
+     */
+    dispose_in_background(std::move(to_evict));
+    if (it == _readers.end()) {
+        vlog(stlog.trace, "{} - reader cache miss for: {}", _ntp, cfg);
+        return std::nullopt;
+    }
+
+    vlog(stlog.trace, "{} - reader cache hit for: {}", _ntp, cfg);
+    it->reader->reset_config(cfg);
+
+    // we use cached_reader wrapper to track reader usage, when cached_reader is
+    // destroyed we unlock reader and trigger eviction
+    return it->make_cached_reader(this);
+}
+
+ss::future<> readers_cache::stop() {
+    co_await _gate.close();
+    for (auto& r : _readers) {
+        co_await r.reader->finally();
```

Comment:
is it really necessary to wait on all the readers.finally? it seems like since cache doesn't ultimately control the lifetime of the user of the reader then this could become problematic if some reader had a long lifetime on shutdown. would an alternative approach work where in the entry_guard you cleared a bit/pointer so that the entyr_guard destructor became a noop? this might be another case where using circular_buffer<lw_shared_ptr> simplifies things.

Reply:
We would wait for all those readers in `disk_log_impl::close()` since it requires write lock.